### PR TITLE
Feature/codeowner in mergebot

### PR DESCRIPTION
**Pull request by SindhuraGowri**
CODEOWNER approval validation against reviewboard reviewers

### CODE REVIEW

File: marge/job.py [hunk 1]
```diff
@@ -516,20 +518,91 @@ def synchronize_using_gitlab_rebase(self, merge_request, expected_sha=None):
...
+
+def get_codeowner_approvers(**params):
+    with open(dirname(dirname(abspath(__file__))) + '/codeowners_approvers.yml',"r") as f:
+        approvers = yaml.load(f, Loader=yaml.FullLoader)
+        if params.get("upgrade_approvers") is not None:
+            return approvers[params.get("upgrade_approvers")]
+        else:
+            return approvers[params.get("api_approvers")]
+
+def codeowner_approval_check(rbt_approvers, merge_request, api):
```
Comment (RajasekharReddyY) [inline marge/job.py:543]: Can we rename to **get_guild_approvers(rbt_approvers, merge_request, api):**

File: marge/job.py [hunk 2]
```diff
@@ -516,20 +518,80 @@ def synchronize_using_gitlab_rebase(self, merge_request, expected_sha=None):
...
+            regexp_file_path = re.compile(file_path)
+            match_files = list(filter(regexp_file_path.match, diff_file_list))
+            log.info('match_files %s', match_files)
+            if match_files:
+                if approvers is not None:
+                    temp = set(approvers) 
+                    codeowners_approvers = [value for value in reviewboard_reviewers if value in temp]
+                    # filtering duplicates
+                    codeowners_approvers = list(dict.fromkeys(codeowners_approvers))
+                    log.info('codeowners_approvers %s', codeowners_approvers)
```
Comment (RajasekharReddyY) [inline marge/job.py:574]: add category in log

File: marge/job.py [hunk 3]
```diff
@@ -51,27 +52,33 @@ def get_approver_branch_data(self, **params):
...
+    def validate_branch_approvers(self, target_branch, rbt_reviewers):
         log.info('%s', target_branch)
         protected_branches = self.get_approver_branch_data(target_branch=None,protected_branches="protected_branches")
         log.info('protected_branches - %s', protected_branches)
-        if target_branch in protected_branches:
-            self.is_Approvers_Eligible(approvers, target_branch)
-        return
+        if target_branch in protected_branches:           
+            approvers = rbt_reviewers.split(',')
+            log.info('approvers... ')
```
Comment (RajasekharReddyY) [inline marge/job.py:77]: Merge both logging

File: marge/job.py [hunk 4]
```diff
@@ -51,27 +52,33 @@ def get_approver_branch_data(self, **params):
...
+        #     valid_appr.remove(user)
+        if not valid_appr:
+            raise CannotMerge('Branch manager approval not found... check branch_approver.yaml file in the project to know the eligible branch managers of this protected branch')        
+        else:
+            approved_by = ','.join([str(elem) for elem in valid_appr])
+            log.info('approved_by: %s', approved_by)
+            return approved_by
+
+    def validate_branch_approvers(self, target_branch, rbt_reviewers):
         log.info('%s', target_branch)
```
Comment (RajasekharReddyY) [inline marge/job.py:72]: Add target branch in log

File: marge/job.py (hunk 3 above)
Comment (SindhuraGowri) [reply marge/job.py:77]: merged

**Pull request merged**